contracts/IndexStaking.sol

```diff
@@ -0,0 +1,61 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.6.6;
+
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "../libraries/SafeTransfer.sol";
+
+contract IndexStaking {
+    using SafeMath for uint256;
+
+    uint public constant duration = 3 * 365 * 6500; // blocks
+    uint public constant totalSupply = 100000000 * 10**18;
+    uint public constant rewardSupply = 48 * totalSupply / 100;
+
+    address public stakingToken;
+    address public rewardToken;
+
+    uint public staked; // S
+    uint public activeStakeDeposits; // T
+    mapping(address => uint) public stake; // stake
+    mapping(address => uint) public stakedSnapshot; // S0
+
+    uint public accumulatedReward;
+    uint public startedBlock;
+
+    constructor(address _stakingTokenAddress, address _rewardTokenAddress) public {
+        stakingToken = _stakingTokenAddress;
+        rewardToken = _rewardTokenAddress;
+        startedBlock = block.number;
+    }
+
+    function deposit(uint _amount) external {
+        SafeTransfer.transferFromERC20(address(stakingToken), msg.sender, address(this), _amount);
+        stake[msg.sender] = _amount;
+        stakedSnapshot[msg.sender] = staked;
+        activeStakeDeposits = activeStakeDeposits.add(_amount);
+    }
+
+    function withdraw() external {
+        uint deposited = stake[msg.sender];
+        uint reward = _calculateReward();
+        if (activeStakeDeposits != 0) {
+            staked = staked.add(reward.div(activeStakeDeposits));
+        }
+        uint userReward = deposited.mul(staked.sub(stakedSnapshot[msg.sender]));
+        activeStakeDeposits = activeStakeDeposits.sub(deposited);
+        stake[msg.sender] = 0;
+        SafeTransfer.sendERC20(address(stakingToken), msg.sender, deposited);
+        SafeTransfer.sendERC20(address(rewardToken), msg.sender, userReward);
+    }
+
+    function _calculateReward() private returns (uint reward) {
+        if (startedBlock.add(duration) > block.number) {
+            reward = rewardSupply
+                .mul(block.number)
+                .mul(block.number.sub(startedBlock))
+                .div((startedBlock.add(duration)).mul(duration));
```

Comment:
1. Change `startedBlock` to `startBlock`. The later is grammatically correct.
2. The formula is not correct. The exact formula should look like:
reward = 0.48 * totalSupply * N * (N - N1) / ((N1 + duration) * duration)
One needs to divide by `duration`, not multiply.

Reply:
2. Multiplication is used in the denominator, (N1 + duration) * duration = `(startedBlock.add(duration)).mul(duration)`  and then it is used to division

Reply:
Ok, that is right.